### PR TITLE
REF: share Index.union

### PR DESCRIPTION
**Pull request by jbrockmendel**

### CODE REVIEW

File: pandas/core/indexes/multi.py [hunk 1]
```diff
@@ -3502,98 +3502,9 @@ def equal_levels(self, other) -> bool:
     # --------------------------------------------------------------------
     # Set Methods
 
-    def union(self, other, sort=None):
```
Comment (jreback) [inline pandas/core/indexes/multi.py:3505]: can you add these examples to the existing Index.union doc-string?

File: pandas/core/indexes/base.py [hunk 2]
```diff
@@ -2656,12 +2656,59 @@ def union(self, other, sort=None):
...
         """
         self._validate_sort_keyword(sort)
         self._assert_can_do_setop(other)
         other, result_name = self._convert_can_do_setop(other)
 
         if not is_dtype_equal(self.dtype, other.dtype):
+            if isinstance(self, ABCMultiIndex) and not is_object_dtype(
+                unpack_nested_dtype(other)
+            ):
+                raise NotImplementedError(
```
Comment (jreback) [inline pandas/core/indexes/base.py:2707]: hit in tests?

File: pandas/core/indexes/base.py [hunk 3]
```diff
@@ -2680,6 +2727,14 @@ def union(self, other, sort=None):
             right = other.astype(dtype, copy=False)
             return left.union(right, sort=sort)
 
+        elif not len(other) or self.equals(other):
```
Comment (jreback) [inline pandas/core/indexes/base.py:2730]: coverage for all of these?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexes/base.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/core/indexes/base.py:2730]: yes

File: pandas/core/indexes/base.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/indexes/base.py:2707]: yes

**Review by jreback — APPROVED** [review]

File: pandas/core/indexes/multi.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/indexes/multi.py:3505]: restored + greenish

**Comment by jreback** [comment]
thanks would be super helpful to either disable / fix the failing travis build (it's a locale one i think)

**Pull request merged**